File: src/ops/crawl.ts
/**
 * wsc crawl — Firecrawl-only with apply gates. No fallback (fail loud).
 *
 * Apply ladder:
 *   ≤10 pages   → no --apply required
 *   11–100      → require --apply
 *   >100        → require --apply --i-know-this-burns-credits
 */

import { withCall } from "../audit.js";
import { queryFingerprint } from "./_chain.js";
import { DisabledError, FirecrawlProvider, MissingKeyError, ProviderError, getProvider } from "../providers/index.js";

export interface CrawlOptions {
  maxPages?: number;
  includePaths?: string[];
  excludePaths?: string[];
  formats?: string[];
  apply?: boolean;
  deepApply?: boolean;
  correlationId?: string;
  noReceipt?: boolean;
}

export function gatePages(maxPages: number, opts: { apply: boolean; deepApply: boolean }): string | null {
  if (maxPages <= 10) return null;
  if (maxPages <= 100 && !opts.apply) return `crawl of ${maxPages} pages requires --apply (range 11–100)`;
  if (maxPages > 100 && !(opts.apply && opts.deepApply)) {
    return `crawl of ${maxPages} pages requires --apply --i-know-this-burns-credits`;
  }
  return null;
}

export async function run(url: string, opts: CrawlOptions = {}): Promise<Record<string, unknown>> {
  const maxPages = opts.maxPages ?? 10;
  const block = gatePages(maxPages, { apply: !!opts.apply, deepApply: !!opts.deepApply });
  if (block) {
    return {
      ok: false,
      operation: "crawl",
      error: block,
      url,
      max_pages: maxPages,
      fallback_chain: [],
      returncode: 2,
    };
  }

  return await withCall(
    "crawl",
    { provider: "firecrawl", correlationId: opts.correlationId, noReceipt: opts.noReceipt },
    async (receipt) => {
      Object.assign(receipt, queryFingerprint(url));
      receipt.params = {
        max_pages: maxPages,
        include_paths: opts.includePaths ?? null,
        exclude_paths: opts.excludePaths ?? null,
        formats: opts.formats ?? null,
        apply: !!opts.apply,
      };
      let provider: FirecrawlProvider;
      try {
        provider = getProvider("firecrawl") as FirecrawlProvider;
      } catch (err) {
        if (err instanceof MissingKeyError || err instanceof DisabledError) {
          receipt.status = "error";
          receipt.fallback_chain = [{ from: "firecrawl", reason: err.kind }];
          return {
            ok: false,
            operation: "crawl",
            provider: null,
            fallback_chain: receipt.fallback_chain,
            error: err.message,
            returncode: 2,
          };
        }
        throw err;
      }
      try {
        const pages = await provider.crawl(url, {
          limit: maxPages,
          includePaths: opts.includePaths,
          excludePaths: opts.excludePaths,
          formats: opts.formats,
        });
        receipt.provider = "firecrawl";
        receipt.fallback_chain = [];
        const urls = pages.map((p) => p.url).filter((u) => !!u);
        receipt.selected_urls = urls;
        receipt.selected_count = urls.length;
        receipt.results_count = pages.length;
        return {
          ok: true,
          operation: "crawl",
          provider: "firecrawl",
          url,
          max_pages: maxPages,
          pages: pages.map((p) => p.toJSON()),
          returncode: 0,
        };
      } catch (err) {
        receipt.status = "error";
        receipt.error = (err instanceof Error ? err.message : String(err)).slice(0, 200);
        return {
          ok: false,
          operation: "crawl",
          provider: "firecrawl",
          error: err instanceof ProviderError ? err.message : String(err),
          returncode: 1,
        };
      }
    },
  );
}

File: cli/src/wsc/ops/crawl.py
"""wsc crawl — Firecrawl crawl with per-call and per-day budget gates.

Apply ladder (locked in by the plan):

* ≤10 pages   → no ``--apply`` required.
* 11–100      → require ``--apply``.
* >100        → require ``--apply --i-know-this-burns-credits``.

There is **no fallback** for crawl — Firecrawl owns this lane. If Firecrawl
is missing/disabled, fail loud.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

from wsc import audit
from wsc.ops._chain import filtered_chain, query_fingerprint
from wsc.providers import (
    DisabledError,
    MissingKeyError,
    ProviderError,
    get_provider,
)


def gate_pages(max_pages: int, *, apply: bool, deep_apply: bool) -> Optional[str]:
    if max_pages <= 10:
        return None
    if max_pages <= 100 and not apply:
        return f"crawl of {max_pages} pages requires --apply (range 11–100)"
    if max_pages > 100 and not (apply and deep_apply):
        return (
            f"crawl of {max_pages} pages requires --apply --i-know-this-burns-credits"
        )
    return None


def run(
    url: str,
    *,
    max_pages: int = 10,
    include_paths: Optional[List[str]] = None,
    exclude_paths: Optional[List[str]] = None,
    formats: Optional[List[str]] = None,
    apply: bool = False,
    deep_apply: bool = False,
    correlation_id: Optional[str] = None,
    no_receipt: bool = False,
) -> Dict[str, Any]:
    block_reason = gate_pages(max_pages, apply=apply, deep_apply=deep_apply)
    if block_reason:
        return {
            "ok": False,
            "operation": "crawl",
            "error": block_reason,
            "url": url,
            "max_pages": max_pages,
            "fallback_chain": [],
            "returncode": 2,
        }

    with audit.start_call(
        "crawl", provider="firecrawl",
        correlation_id=correlation_id, no_receipt=no_receipt,
    ) as receipt:
        receipt.update(query_fingerprint(url))
        receipt["params"] = {
            "max_pages": max_pages,
            "include_paths": include_paths,
            "exclude_paths": exclude_paths,
            "formats": formats,
            "apply": apply,
        }
        try:
            provider = get_provider("firecrawl")
        except (MissingKeyError, DisabledError) as exc:
            receipt["status"] = "error"
            receipt["fallback_chain"] = [{"from": "firecrawl", "reason": exc.kind}]
            return {
                "ok": False,
                "operation": "crawl",
                "provider": None,
                "fallback_chain": receipt["fallback_chain"],
                "error": str(exc),
                "returncode": 2,
            }
        try:
            pages = provider.crawl(
                url,
                limit=max_pages,
                include_paths=include_paths,
                exclude_paths=exclude_paths,
                formats=formats,
            )
        except ProviderError as exc:
            receipt["status"] = "error"
            receipt["error"] = str(exc)[:200]
            return {
                "ok": False,
                "operation": "crawl",
                "provider": "firecrawl",
                "error": str(exc),
                "returncode": 1,
            }
        receipt["provider"] = "firecrawl"
        receipt["fallback_chain"] = []
        urls = [p.url for p in pages if p.url]
        receipt["selected_urls"] = urls
        receipt["selected_count"] = len(urls)
        receipt["results_count"] = len(pages)
        return {
            "ok": True,
            "operation": "crawl",
            "provider": "firecrawl",
            "url": url,
            "max_pages": max_pages,
            "pages": [p.to_dict() for p in pages],
            "returncode": 0,
        }
